lib/instrumentation/async-hooks.js

```diff
@@ -0,0 +1,35 @@
+'use strict'
+
+const asyncHooks = require('async_hooks')
+
+module.exports = function (ins) {
+  const asyncHook = asyncHooks.createHook({init, destroy})
+  const transactions = new Map()
+
+  Object.defineProperty(ins, 'currentTransaction', {
+    get: function () {
+      const asyncId = asyncHooks.executionAsyncId()
+      return transactions.get(asyncId)
+    },
+    set: function (trans) {
+      const asyncId = asyncHooks.executionAsyncId()
+      transactions.set(asyncId, trans)
+    }
+  })
+
+  asyncHook.enable()
+
+  function init (asyncId, type, triggerAsyncId, resource) {
+    // We don't care about the TIMERWRAP, as it will only init once for each
+    // timer that shares the timeout value. Instead we rely on the Timeout
+    // type, which will init for each scheduled timer.
+    if (type === 'TIMERWRAP') return
+
+    transactions.set(asyncId, ins.currentTransaction)
+  }
+
+  function destroy (asyncId) {
+    if (!transactions.has(asyncId)) return // in case type === TIMERWRAP
```

Comment:
I imagine leaving this up to the `Map.prototype.delete` function is just as efficient

Reply:
Same as `get`, this map deopt if it tries to delete an item that doesn't exist.